subsys/fs/nvs/nvs.c

```diff
@@ -319,37 +320,30 @@ static int _nvs_prev_ate(struct nvs_fs *fs, u32_t *addr, struct nvs_ate *ate)
 		return rc;
 	}
 
-	while (1) {
-		*addr += ate_size;
-		if (((*addr) & ADDR_OFFS_MASK) != fs->sector_size) {
-			break;
-		}
-		/* last ate in sector, do jump to previous sector */
-		if (((*addr) >> ADDR_SECT_SHIFT) == 0) {
-			*addr += (fs->sector_count << ADDR_SECT_SHIFT);
-		}
-		(*addr) -= (1 << ADDR_SECT_SHIFT);
-		while (1) {
-			*addr -= ate_size;
-			rc = _nvs_flash_cmp_const(fs, *addr, 0xff, sizeof(struct nvs_ate));
-			if (!rc) {
-				break;
-			}
-			rc = _nvs_flash_cmp_const(fs, *addr, 0x00, sizeof(struct nvs_ate));
-			if (!rc) {
-				break;
-			}
-			/* stop at the end of the sector in case it is not
-			   closed properly */
-			if (((*addr) & ADDR_OFFS_MASK) == 0) {
-				break;
-			}
-		}
-		/* stop after stepping through all sectors */
-		if (*addr == fs->ate_wra) {
-			break;
-		}
+	*addr += ate_size;
+	if (((*addr) & ADDR_OFFS_MASK) != (fs->sector_size - ate_size)) {
+		return 0;
+	}
+
+	/* last ate in sector, do jump to previous sector */
+	if (((*addr) >> ADDR_SECT_SHIFT) == 0) {
+		*addr += (fs->sector_count << ADDR_SECT_SHIFT);
 	}
+	(*addr) -= (1 << ADDR_SECT_SHIFT);
+
+	rc = _nvs_flash_cmp_const(fs, *addr, 0xff, sizeof(struct nvs_ate));
+	/* at the end of filesystem */
+	if (!rc) {
+		*addr = fs->ate_wra;
+		return 0;
+	}
+
+	rc = _nvs_flash_ate_rd(fs, *addr, &close_ate);
+	if (rc) {
+		return rc;
+	}
+	(*addr) &= ADDR_SECT_MASK;
+	(*addr) += close_ate.offset;
```

Comment:
here the clue of close-ate

Reply:
I don't get what you mean, maybe I should add some comment:
/* update the address so it points to the last added ate */

Reply:
yes, pleas do